### ISSUE TITLE
Istio 1.1.1 breaks routing

### ISSUE THREAD

**Issue by emedina** [istio/istio#12801]
**Describe the bug**
After upgrading from `1.1.0` to `1.1.1`, all the routing stopped working. Everything responds with a `404`.

**Expected behavior**
Routes should not be impacted by a minor revision upgrade!

**Steps to reproduce the bug**
Upgrade from an existing `1.1.0` with some routes (`gateway` + `virtualservice`) and then upgrade to `1.1.1` and you should get `404`.

**Version**
Working with `1.1.0`
Everything `404` with `1.1.1`

**Installation**
Helm chart

**Environment**
Kubernetes 1.13

**Comment by ryan-dyer** [comment]
I am seeing the same thing with the 1.1.1 release.

**Comment by Stono** [comment]
So i didn't see this and wondering what is different! 

I don't use `Gateway` at all, but have 200+ `VirtualService`

**Comment by hgontijo** [comment]
After upgrading from `1.1.0` to `1.1.1`, I'm facing similar issue.
Routing on `:80` is working fine, however it's failing for `:443`:
```
[2019-03-26 22:46:14.020][52][warning][config] [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_mux_subscription_lib/common/config/grpc_mux_subscription_impl.h:70]
gRPC config for type.googleapis.com/envoy.api.v2.Listener rejected:
Error adding/updating listener 0.0.0.0_443: error adding listener '0.0.0.0:443': multiple filter chains with the same matching rules are defined
```

**Comment by howardjohn** [comment]
Just looking at the release notes, it seems like this is most likely related to "Add namespace scoping to the Gateway ‘port’ names." @philrud

or possibly " Do not apply VirtualService rule to the wrong gateway" @rshriram

The other changes seem unlikely to be related

**Comment by philrud** [comment]
@emedina @hgontijo I wasn't able to reproduce it yet with a couple of rather trivial examples that I've tried. Could you please provide some extra details so we know what to look for? Specifically:

* please share some yamls -- at least your gateway/virtualservice definitions
* please provide installation options/flags that were used to install 1.1.0/upgrade to 1.1.1
* is this only happening as the result of an 1.1.0 -> 1.1.1 upgrade? did you have a chance to try reproducing this with a clean 1.1.1 installation?
* is it permanently broken with 1.1.1 or would it recover eventually after upgrade is finished?

thanks!

**Comment by philrud** [comment]
Okay, looks like I'm able to reproduce 404s during a short time window right after 1.1.0 -> 1.1.1 upgrade (eventually they go away in my case though, would be great if anybody who's hitting this issue could confirm whether the 404s are permanent).

This seems to be related to my changeset: https://github.com/istio/istio/pull/12556

**Comment by emedina** [comment]
@philrud Please find here the answers to your questions:

* please share some yamls -- at least your gateway/virtualservice definitions
_https://gist.github.com/emedina/ba73a409158d8f0bfbfd7c60bfdd4240_

* please provide installation options/flags that were used to install 1.1.0/upgrade to 1.1.1
is this only happening as the result of an 1.1.0 -> 1.1.1 upgrade? did you have a chance to try reproducing this with a clean 1.1.1 installation?
_Reproducible for me as a clean 1.1.1 installation (deleted ìstio-system` namespace`, `crd`, then re-install everything from scratch)._

* is it permanently broken with 1.1.1 or would it recover eventually after upgrade is finished?
_Permanently broken, although I do see the `clusters` and `listeners` in the `Ingress gateway proxy`._

Let me know where I can help further.

**Comment by hpohl** [comment]
I am also getting 404s, but they appeared already shortly after upgrading to 1.1.0 from 1.0.6. Upgrading to 1.1.1 did not change anything. Trying to reproduce it locally on a minikube.

**Comment by philrud** [comment]
Thanks, @emedina. 

So this definitely has nothing to do with #12556 as that change only affects HTTPS routes and all of @emedina's gateways are HTTP. I've tried that setup fitted with a few helloworld-like services -- that actually works for me, I can't reproduce the 404s. 

Also looks like 404s indeed aren't new as @hpohl mentioned. I can reproduce the same spurious 404s during 1.0.x -> 1.1.0 upgrade as well, so I'm taking back what I said earlier regarding that it may be related to #12556 (worth noting I was trying with only a single ingressgateway/single pilot instance).

If anybody can come up with an isolated example and steps to reproduce the issue (with either https or http case) that would be super-useful.

**Comment by phanama** [comment]
Also happened to me, but only on GPRC/HTTP2 protocols

**Comment by hpohl** [comment]
Here is a gist of a slimmed down setup triggering this issue: https://gist.github.com/hpohl/85a2195ff798845ae38821446136c9b9

One of the services works while the other one just returns 404s. Which one is working and which one not seems to be picked by random.

**Comment by howardjohn** [comment]
@hpohl I think this is the same issue as https://github.com/istio/istio/issues/13014

**Comment by hpohl** [comment]
Yes, can confirm it's that issue. So maybe this one is a duplicate (or the other one).

**Comment by philrud** [comment]
Closing this due to lack of information. If there's any additional information on how to reproduce this, please re-open or rather open a more specific issue.

**Comment by howardjohn** [comment]
@philrud they provided yaml of their setup to reproduce. what other info is needed? I don't think this should be closed

**Comment by philrud** [comment]
@howardjohn oh, so we're not yet sure if this is the port conflict? I've misinterpreted your prev. comment.

@hpohl's issue isn't 1.1.1 specific according to his prev. comments so probably should be tracked separately (or as part of #13014 if we believe this is a dup) as this is getting very confusing.

**Comment by howardjohn** [comment]
I don't think the original issue is the same, but looking closer it is unclear, maybe they are a dupe. @emedina do you have two services on the same port targeting different container ports?

**Comment by philrud** [comment]
@howardjohn yeah, I get it now. Regarding @emedina case, see the conversation above, here's the yaml: https://gist.github.com/emedina/ba73a409158d8f0bfbfd7c60bfdd4240 – I've tried it with 1.1.1, that actually works (after substituting container images with a helloworld-style app), I wasn't able to reproduce the behavior that was described

**Comment by emedina** [comment]
@howardjohn Yes, the port where the Pod's API is listening and the port where the Pod's health is listening. My Pods are typically Spring Boot applications where the API listens on 8080 and health actuator on 8081.

**Comment by emedina** [comment]
It seems I'm still facing this issue with `1.1.3`, but I will confirm in a while whether it's actually the case. I will downgrade to `1.0.1` and see...

**Comment by wolfgang-braun** [comment]
@emedina do you still face this issue?

**Comment by emedina** [comment]
This has been solved at least in `1.1.8`, so closing the issue.
